src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py

```diff
@@ -0,0 +1,81 @@
+import torch
+import torch.nn as nn
+
+
+class MaxOut(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, inputs):
+        current_max = inputs[0]
+        for inp in inputs[1:]:
+            current_max = torch.max(current_max, inp)
+        return current_max
+
+
+class ConvCompetitiveLayer(nn.Module):
+
+    def __init__(self, in_channels, nb_kernels, kernels_sizes):
+        super(ConvCompetitiveLayer, self).__init__()
+
+        self.convs = [nn.Conv2d(in_channels, nb_kernels, kernel_size, padding=kernel_size//2)
+                      for kernel_size in kernels_sizes]
+
+        self.bnorms = [nn.BatchNorm2d(nb_kernels) for _ in kernels_sizes]
+        self.maxout = MaxOut()
+
+    def forward(self, x):
+        outs = [bnorm(conv(x)) for conv, bnorm in zip(self.convs, self.bnorms)]
+        return self.maxout(outs)
+
+
+class CompetitiveMultiscaleCNN(nn.Module):
+    """Implementation of Multiscale Competitive CNN proposed in:
+        https://arxiv.org/abs/1511.05635
+    """
+
+    def __init__(self, in_channels, avg_pool_size=3):
+        """Create MultiscaleCompetitiveCNN object accepting images with given number of channels.
+
+        :param in_channels: Number of channels in the input image. Image size is irrelevant.
+        :param avg_pool_size: Pool size for the last layer - Average Pooling
+            (see https://arxiv.org/abs/1511.05635 - pool size not specified in the paper).
+        """
+        super(CompetitiveMultiscaleCNN, self).__init__()
+
+        self.block0_layer_0 = ConvCompetitiveLayer(in_channels, 192, [1, 3, 5, 7])
+        self.block0_layer_1 = ConvCompetitiveLayer(192, 160, [1, 3])
+        self.block0_layer_2 = ConvCompetitiveLayer(160, 96, [1, 3])
+
+        self.block0_maxpool = nn.MaxPool2d(3)
+        self.block0_dropout = nn.Dropout2d()
+
+        self.block1_layer_0 = ConvCompetitiveLayer(96, 192, [1, 3, 5, 7])
+        self.block1_layer_1 = ConvCompetitiveLayer(192, 192, [1, 3])
+        self.block1_layer_2 = ConvCompetitiveLayer(192, 192, [1, 3])
+
+        self.block1_maxpool = nn.MaxPool2d(3)
+        self.block1_dropout = nn.Dropout2d()
+
+        self.block2_layer_0 = ConvCompetitiveLayer(192, 192, [1, 3, 5, 7])
+        self.block2_layer_1 = ConvCompetitiveLayer(192, 192, [1, 3])
+        self.block2_layer_2 = ConvCompetitiveLayer(192, 10, [1, 3])
+
+        self.block2_avgpool = nn.AvgPool2d(avg_pool_size)
```

Comment:
Still non adaptive pooling size. Now it requires from user to know the implementation of the model so he can put appropriate value in the constructor.